Environments + client services w/prerendering

guardrex opened the pull request:
Fixes #31025 
Addresses #28161

Mackinnon ... I decided to draft this with a full explanation and examples for what's going on with client-side services in a BWA. It's a little long. However, I can understand it this way, so we have a good bet on not losing anyone.

In the *Render Modes* article, the *Set the render mode for the entire app* section was too far down the article. I moved the section up and changed the heading name to *Apply a render mode to the entire app*. The new section on *Client-side services fail to resolve during prerendering* makes reference to globally setting the render mode. I want to reader to have already seen that section.

<!-- PREVIEW-TABLE-START -->

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/blazor/components/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/76f68fd99e58275e5ea18a1e16d15593dd1c196a/aspnetcore/blazor/components/index.md) | [ASP.NET Core Razor components](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/components/index?branch=pr-en-us-31057) |
| [aspnetcore/blazor/components/render-modes.md](https://github.com/dotnet/AspNetCore.Docs/blob/76f68fd99e58275e5ea18a1e16d15593dd1c196a/aspnetcore/blazor/components/render-modes.md) | [ASP.NET Core Blazor render modes](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/components/render-modes?branch=pr-en-us-31057) |
| [aspnetcore/blazor/fundamentals/configuration.md](https://github.com/dotnet/AspNetCore.Docs/blob/76f68fd99e58275e5ea18a1e16d15593dd1c196a/aspnetcore/blazor/fundamentals/configuration.md) | [ASP.NET Core Blazor configuration](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/configuration?branch=pr-en-us-31057) |
| [aspnetcore/blazor/fundamentals/environments.md](https://github.com/dotnet/AspNetCore.Docs/blob/76f68fd99e58275e5ea18a1e16d15593dd1c196a/aspnetcore/blazor/fundamentals/environments.md) | [aspnetcore/blazor/fundamentals/environments](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/environments?branch=pr-en-us-31057) |
| [aspnetcore/blazor/security/server/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/76f68fd99e58275e5ea18a1e16d15593dd1c196a/aspnetcore/blazor/security/server/index.md) | [Secure ASP.NET Core server-side Blazor apps](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/security/server/index?branch=pr-en-us-31057) |
| [aspnetcore/blazor/state-management.md](https://github.com/dotnet/AspNetCore.Docs/blob/76f68fd99e58275e5ea18a1e16d15593dd1c196a/aspnetcore/blazor/state-management.md) | [ASP.NET Core Blazor state management](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/state-management?branch=pr-en-us-31057) |


<!-- PREVIEW-TABLE-END -->

MackinnonBuck commented on aspnetcore/blazor/fundamentals/environments.md:
````diff
@@ -135,57 +152,208 @@ In the following example for IIS, the custom header (`Blazor-Environment`) is ad
+In the server project's `Program` file, register a `IBlazorHostEnvironment` service with the `BlazorHostEnvironment` implementation. Set the host environment with <xref:Microsoft.AspNetCore.Builder.WebApplicationBuilder.Environment%2A?displayProperty=nameWithType>, which is only available on the server:
+
+```csharp
+builder.Services.AddSingleton<IBlazorHostEnvironment>(
+    sp => new BlazorHostEnvironment(builder.Environment.EnvironmentName));
+```
+
+In the `.Client` project's `Program` file, register a `IBlazorHostEnvironment` service with the `BlazorHostEnvironment` implementation. Set the host environment with <xref:Microsoft.AspNetCore.Components.WebAssembly.Hosting.IWebAssemblyHostEnvironment.Environment%2A?displayProperty=nameWithType>, which is only available on the client:
+
+```csharp
+builder.Services.AddSingleton<IBlazorHostEnvironment>(
+    sp => new BlazorHostEnvironment(builder.HostEnvironment.Environment));
````
This is definitely valid, albeit different than what I had in mind.

If the customer were to add additional members to the `IBlazorHostEnvironment` interface, it would require adding more parameters to the primary constructor and providing values in `Program.cs`. Or if the interface had any methods in it, it would require falling back to separate implementations (unless you do a thing where you dynamically check which platform you're on using `OperatingSystem.IsBrowser()`).

I think it would be a bit more conventional to have separate implementations of `IBlazorHostEnvironment` that can inject app-specific dependencies and use those to support aspects of the implementation. For example:

`.Client` project:
```csharp
// IBlazorHostEnvironment.cs
public interface IBlazorHostEnvironment
{
    string EnvironmentName { get; }
}

// BlazorWebAssemblyHostEnvironment.cs
public class BlazorWebAssemblyHostEnvironment(IWebAssemblyHostEnvironment hostEnvironment) : IBlazorHostEnvironment
{
    public string EnvironmentName => hostEnvironment.Environment;
}

// Program.cs
builder.Services.AddSingleton<IBlazorHostEnvironment, BlazorWebAssemblyHostEnvironment>();
```

Server project:
```csharp
// BlazorServerHostEnvironment.cs
public class BlazorServerHostEnvironment(IWebHostEnvironment hostEnvironment) : IBlazorHostEnvironment
{
    public string EnvironmentName => hostEnvironment.EnvironmentName;
}

// Program.cs
builder.Services.AddScoped<IBlazorHostEnvironment, BlazorServerHostEnvironment>();
```

What do you think?

guardrex replied:
I did it that way because I felt there probably wouldn't be anything else, but I think the most general approach is what you're suggesting. Since we'll use this as a promonient example of how to address this problem, I think your idea is best.

WRT the dispose situation from above. I'm confused if the DI container is creating the `new`-ed up services like this. I thought that it was NOT creating them ... that the dev code is creating them. If that's right, then the disposal code is boilerplate code that doesn't dispose of anything but perhaps present as a good pattern so as to be clear that the service implementation needs to dispose of anything it creates ... just as a reminder ... a pattern. Alternatively if I'm incorrect about this service registration and what's actually happening is that the DI container is creating them, then yeah ... for sure there's no need to place this boilerplate do-nothing disposal code ... such as what we'll be changing over to when I adopt your suggestion here for the serivces.

https://learn.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-8.0#disposal-of-services

MackinnonBuck commented on aspnetcore/blazor/components/render-modes.md:
```diff
@@ -284,6 +312,43 @@ In the following example, the component is interactive throughout the process. T
+
+* *Recommended*: Create a service abstraction and create implementations&dagger; for the service in the `.Client` and server projects. Register the services in each project. Inject the service in the component. For a demonstration of this approach, see <xref:blazor/fundamentals/environments#read-the-client-side-environment-in-a-blazor-web-app>.
+
+  &dagger;One implementation can be created if the implementation can be configured for execution in the `.Client` and server projects.
```
Update: I now see what you mean by looking at your example - I left a comment down there 🙂

guardrex replied:
I'll fix this on the next commit when I update the example approach.

MackinnonBuck reviewed (commented):
Looks like a great start - I have some ideas and thoughts!

guardrex commented on aspnetcore/blazor/components/render-modes.md:
````diff
@@ -142,6 +142,34 @@ Technically, `@rendermode` is both a Razor *directive* and a Razor *directive at
+<Routes @rendermode="InteractiveServer" />
+```
+
+The Blazor router propagates its render mode to the pages it routes. The pages aren't technically child components of the router, but when the routes are discovered at runtime for the router, they inherit the router's render mode.
````
It's probably ok to 🔪 that. IIRC, that was one of Dan's remarks.

guardrex commented on aspnetcore/blazor/components/render-modes.md:
````diff
@@ -284,6 +312,43 @@ In the following example, the component is interactive throughout the process. T
+
+```razor
+@page "/"
+@inject IBlazorHostEnvironment Environment
````
Yes ... I was going toooooo fast. Normally, I'd sleep on something this big and fix 😈 bugs in it on the next day.

guardrex commented on aspnetcore/blazor/components/render-modes.md:
```diff
@@ -284,6 +312,43 @@ In the following example, the component is interactive throughout the process. T
+
+No compile time error occurs, but a runtime error occurs during prerendering:
+
+> :::no-loc text="Cannot provide a value for property 'HostEnvironment' on type 'BlazorWebApp80EnvironmentTesting.Client.Pages.Home'. There is no registered service of type 'Microsoft.AspNetCore.Components.WebAssembly.Hosting.IWebAssemblyHostEnvironment'.":::
+
```
Yes ... that was a late change to the example that I didn't spot here.

guardrex commented on aspnetcore/blazor/components/render-modes.md:
```diff
@@ -284,6 +312,43 @@ In the following example, the component is interactive throughout the process. T
+
+No compile time error occurs, but a runtime error occurs during prerendering:
+
+> :::no-loc text="Cannot provide a value for property 'HostEnvironment' on type 'BlazorWebApp80EnvironmentTesting.Client.Pages.Home'. There is no registered service of type 'Microsoft.AspNetCore.Components.WebAssembly.Hosting.IWebAssemblyHostEnvironment'.":::
```
```suggestion
> :::no-loc text="Cannot provide a value for property 'Environment' on type 'BlazorWebApp80EnvironmentTesting.Client.Pages.Home'. There is no registered service of type 'Microsoft.AspNetCore.Components.WebAssembly.Hosting.IWebAssemblyHostEnvironment'.":::
```

guardrex commented on aspnetcore/blazor/fundamentals/environments.md:
```diff
@@ -135,57 +152,208 @@ In the following example for IIS, the custom header (`Blazor-Environment`) is ad
+* Visual Studio Code/.NET CLI: Execute `dotnet new classlib -o {PROJECT NAME}` from a command prompt in the solution's folder. The `-o|--output` option creates a folder for the class library and names the project. For the `{PROJECT NAME}` placeholder, use the name of the server project with "`.Shared`" added to the end of the name (for example: `BlazorWebAppSample.Shared`).
+
+Create a project reference for the class library in both the server and `.Client` projects:
 
-:::code language="razor" source="~/../blazor-samples/3.1/BlazorSample_WebAssembly/Pages/environments/ReadEnvironment.razor" highlight="3,7":::
+* Visual Studio: Right-click each project and select **Add** > **Project Reference**. Select the checkbox for the shared project and select **OK**.
+* Visual Studio Code/.NET CLI: Execute `dotnet add reference {PATH}` in a command shell from each of the server and `.Client` project folders. The `{PATH}` placeholder is the path to the shared class library project.
+
+In the shared (`.Shared`) project:
+
+* Add an interface for the service abstraction.
+* Add classes for the service implementation and host environment extensions.
```
Yes, I was split in my mind on if that felt good, and that is what you said in the email.

I just narrowly was leaning in the direction of portability, using a class lib or package across BWAs. Also, it feels like we're dealing with two distinct projects that are sharing a central resource, even if that isn't the best way to think about it.

However, it would be simpler to explain that way ... in the `.Client` project, so let's shorten this up with that approach. I'll make the changes.

guardrex commented on aspnetcore/blazor/fundamentals/environments.md:
```diff
@@ -135,57 +152,208 @@ In the following example for IIS, the custom header (`Blazor-Environment`) is ad
+    string EnvironmentName { get; set; }
+}
+
+public class BlazorHostEnvironment(string environment) : IBlazorHostEnvironment, 
+    IDisposable
+{
+    private bool _disposed;
+
+    public string EnvironmentName { get; set; } = environment;
+
+    public void Dispose()
+    {
+        if (_disposed)
+        {
+            return;
+        }
+
+        _disposed = true;
+    }
+}
```
I'll answer on this below.

guardrex commented on aspnetcore/blazor/fundamentals/environments.md:
```diff
@@ -135,57 +152,208 @@ In the following example for IIS, the custom header (`Blazor-Environment`) is ad
+    }
+}
+
+public static class BlazorHostEnvironmentExtensions
+{
+    public static bool IsDevelopment(this IBlazorHostEnvironment hostEnvironment)
+    {
+        return hostEnvironment.IsEnvironment("Development");
+    }
+
+    public static bool IsStaging(this IBlazorHostEnvironment hostEnvironment)
+    {
+        return hostEnvironment.IsEnvironment("Staging");
+    }
+
+    public static bool IsProduction(this IBlazorHostEnvironment hostEnvironment)
+    {
+        return hostEnvironment.IsEnvironment("Production");
+    }
+
+    public static bool IsEnvironment(
+        this IBlazorHostEnvironment hostEnvironment,
+        string environmentName)
+    {
+        return string.Equals(
+            hostEnvironment.EnvironmentName,
+            environmentName,
+            StringComparison.OrdinalIgnoreCase);
+    }
+}
```
Well ... it's kind'a nice to get the methods. I wrote this so that devs can just use it just like this ... save them some time. It's fast to delete the bits they don't want. It's slower (and a pain) for them to have to create these.

guardrex commented:
@MackinnonBuck ... I think I have most of the ....

# 🦖 ***RexHacks!***&trade; 🦖

... out of the PR. 🙈 ... but you can see why I like to get these technical reviews. You never know what kind of disgusting 💩 code a 🦖 will poop out into the doc! 😆

I shortened the API naming quite a bit to get everything in under our 85 character line length for doc code. It doesn't have a collision with `Microsoft.Extensions.Hosting.IHostEnvironment` AFAICT.

guardrex commented:
... and I changed the section name in the *Environments* topic ...

```diff
Read the client-side environment in a Blazor Web App
Read the environment in a Blazor Web App
```

... because the environment is being read in both and this has broader implications because it also applies to Auto components.

MackinnonBuck commented on aspnetcore/blazor/fundamentals/environments.md:
````diff
@@ -135,57 +152,150 @@ In the following example for IIS, the custom header (`Blazor-Environment`) is ad
+`IHostEnvironment.cs`:
+
+```csharp
+public interface IHostEnvironment
````
The `IHostEnvironment` name might clash with [`Microsoft.Extensions.Hosting.IHostEnvironment`](https://source.dot.net/#Microsoft.Extensions.Hosting.Abstractions/IHostEnvironment.cs,37d39d4dc6955525). Do you think we could find another name for it? Maybe `IAppHostEnvironment` or `IBlazorHostEnvironment`?

guardrex replied:
Yes ... I just need to keep it short. It wasn't colliding in my test app, but I agree that this is a bit scary!

MackinnonBuck commented on aspnetcore/blazor/fundamentals/environments.md:
````diff
@@ -135,57 +152,150 @@ In the following example for IIS, the custom header (`Blazor-Environment`) is ad
+```csharp
+public interface IHostEnvironment
+{
+    string EnvironmentName { get; set; }
````
I think we could make this just `{ get; }`

guardrex replied:
I thought IntelliSense was complaining when I tried that, but ***No!*** It's cool. I'll make the update.

MackinnonBuck commented on aspnetcore/blazor/fundamentals/environments.md:
```diff
@@ -135,57 +152,150 @@ In the following example for IIS, the custom header (`Blazor-Environment`) is ad
+
+public class WasmHostEnvironment(IWebAssemblyHostEnvironment env) : IHostEnvironment
+{
+    public string EnvironmentName { get; set; } = env.Environment;
```
Could this be a computed property (using `=>` rather than `{ get; set; } =`)? Same suggestion for `ServerHostEnvironment`. I only suggest this because `IWebHostEnvironment` has a mutable `EnvironmentName` property, and we would want the `EnvironmentName` property here to reflect any changes to the underlying environment name.

guardrex replied:
Yes ... again, it looked like at first that the compiler was going to choke on it. ***No!*** The compiler is good. I'll make the change. 👍

MackinnonBuck reviewed (commented):
Looking good! I do also want to hear an opinion from @javiercn or @SteveSandersonMS before we officially suggest this approach to customers.

guardrex commented:
@MackinnonBuck ... Looks good, but can we use `IAppEnvironment`? Trying to keep it short to keep the code line lengths at 85 or less chars.

MackinnonBuck commented:
> @MackinnonBuck ... Looks good, but can we use `IAppEnvironment`? Trying to keep it short to keep the code line lengths at 85 or less chars.

That sounds fine to me! 🙂

SteveSandersonMS commented on aspnetcore/blazor/components/render-modes.md:
```diff
@@ -284,6 +312,41 @@ In the following example, the component is interactive throughout the process. T
+
+This error occurs because the component must compile and execute on the server during prerendering, but <xref:Microsoft.AspNetCore.Components.WebAssembly.Hosting.IWebAssemblyHostEnvironment> isn't a registered service on the server.
+
+There are a four approaches that you can take to address this scenario. The following are listed from most recommended to least recommended:
```
Often a simpler option is to inject `IServiceProvider` instead of the service type itself, and then you can have code like:

```cs
if (Services.GetService<IWebAssemblyHostEnvironment>() is {} env)
{
   // ... do stuff with env
}
```

TBH that seems easier than making a new abstraction.

guardrex replied:
AH! I see ... yes ... ***BUT***, it won't resolve an environment at all during prerendering. If the dev is counting on the value there, this will 💥 break their logic, correct?

SteveSandersonMS commented on aspnetcore/blazor/fundamentals/environments.md:
```diff
@@ -39,33 +36,41 @@ For general guidance on ASP.NET Core app configuration, see <xref:fundamentals/e
+                
+We recommend the following conventions:
+
+* Always use the "`Development`" environment name for local development. This is because the Blazor framework expects exactly that name when configuring the app and tooling for local development runs of a Blazor app.
```
It's ASP.NET Core as a whole that's coupled to the name "Development", not Blazor specifically.

SteveSandersonMS commented:
@guardrex @MackinnonBuck Could you clarify which specific part of this doc is the "approach" that you consider at risk of being controversial, and are asking for feedback about? I've read through it and nothing instantly springs out as being problematic, but it is a long doc and maybe I've missed something.

SteveSandersonMS commented:
Oh wait, I think I see now. You're referring to the stuff about `IAppEnvironment`.

While this approach would certainly work, it has the drawback that it doesn't create compatibility with existing code that already depends on accessing `IWebAssemblyHostEnvironment`. Two other possibilities:

## Implement IWebAssemblyHostEnvironment on the server

You may have considered this, not sure, but why not simply create your own service implementation for `IWebAssemblyHostEnvironment` on the server?

```cs
public class ServerHostEnvironment(IWebHostEnvironment env, NavigationManager nav) : IWebAssemblyHostEnvironment
{
    public string Environment => env.EnvironmentName;
    public string BaseAddress => nav.BaseUri;
}
```

... with `builder.Services.TryRegisterScoped<IWebHostEnvironment, ServerHostEnvironment>()`. Now all existing code will just work.

## Fix it in the framework

Longer term, if this is a common pain point, we could change the framework code that registers the WebAssembly hosting endpoints etc so that it also registers an implementation of `IWebAssemblyHostEnvironment` just like the one above, so then the whole problem vanishes for everyone. If this seems right (@MackinnonBuck?) we should track an issue for it.

javiercn commented on aspnetcore/blazor/components/render-modes.md:
```diff
@@ -142,6 +142,34 @@ Technically, `@rendermode` is both a Razor *directive* and a Razor *directive at
 
+## Apply a render mode to the entire app
+
+To set the render mode for the entire app, indicate the render mode at the highest-level interactive component in the app's component hierarchy that isn't a root component.
```
A concrete example here will be good. It's impossible for anyone without context to understand what is being said here. (To give you an idea, I have context and I had to stop to think about it).

guardrex replied:
The definition of a render mode and how to set it for component instances and component definitions were just explained in the preceding sections. What's not clear?

javiercn replied:
I did not go through the previous section. Let me read through it

guardrex commented:
> create your own service implementation for IWebAssemblyHostEnvironment on the server?

I like it, but I recommend keeping the service-based approach listed in case it has utility in other cases.

One thing tho ... is `TryRegisterScoped` valid API? Did you mean `TryAddScoped`?

WRT your earlier remark ...

> ```csharp
> if (Services.GetService<IWebAssemblyHostEnvironment>() is {} env)
> {
>    // ... do stuff with env
> }
> ```

I need more information to understand what problem that's addressing in this context.

guardrex commented on aspnetcore/blazor/components/render-modes.md:
```diff
@@ -142,6 +142,34 @@ Technically, `@rendermode` is both a Razor *directive* and a Razor *directive at
+To set the render mode for the entire app, indicate the render mode at the highest-level interactive component in the app's component hierarchy that isn't a root component.
+
+> [!NOTE]
+> Making a root component interactive, such as the `App` component, isn't supported because the Blazor script may be evaluated multiple times. Therefore, the render mode for the entire app can't be set directly by the `App` component.
```
```suggestion
> Making a root component interactive, such as the `App` component, isn't supported. Therefore, the render mode for the entire app can't be set directly by the `App` component.
```

guardrex commented:
@SteveSandersonMS ... I placed your approach as the top pick. I just moved the ***concept*** of custom services down to the second spot in the list, and it won't have an example in the docs. In the *Environments* article, I placed the full example that you provided.

SteveSandersonMS commented on aspnetcore/blazor/components/render-modes.md:
```diff
@@ -142,6 +142,34 @@ Technically, `@rendermode` is both a Razor *directive* and a Razor *directive at
+To set the render mode for the entire app, indicate the render mode at the highest-level interactive component in the app's component hierarchy that isn't a root component.
+
+> [!NOTE]
+> Making a root component interactive, such as the `App` component, isn't supported because the Blazor script may be evaluated multiple times. Therefore, the render mode for the entire app can't be set directly by the `App` component.
```
In my recollection, it was that the script was being evaluated multiple times.

Nonetheless we don't have to explain *why* it's unsupported for the root component to be interactive. We only need to communicate that it is unsupported.

guardrex replied:
I fixed it by removing the reason.

SteveSandersonMS commented:
> One thing tho ... is TryRegisterScoped valid API? Did you mean TryAddScoped?

Yeah I probably did mean TryAddScoped. Sorry. I was just typing from memory and haven't verified that the code is correct and would compile.

guardrex commented:
I added the `IServiceProvider` approach as a approach, but I did so above the list of *complete* solutions to the problem ... i.e., it isn't useful for obtaining a value during prerendering (if that's what the dev's spec calls for).

@MackinnonBuck, your bullet number 2 was @SteveSandersonMS's idea in code form. I removed that bullet from the list given that I'm moving it up above the list.

I also moved your last bullet on disabling prerendering to just under the `IServiceProvider` approach because that also doesn't solve the problem where the dev wants prerendering and a value.

The list is meant to include only the approaches that fully solve the problem, which is to obtain the value during prerendering. I think the list has the three good approaches.

I recommend re-reviewing the entire section along with the new example in the *Environments* doc.

javiercn commented on aspnetcore/blazor/fundamentals/environments.md:
```diff
@@ -135,57 +152,102 @@ In the following example for IIS, the custom header (`Blazor-Environment`) is ad
 > [!NOTE]
 > To use a custom `web.config` file for IIS that isn't overwritten when the app is published to the `publish` folder, see <xref:blazor/host-and-deploy/webassembly#use-a-custom-webconfig>.
+>
+> Although the Blazor framework issues the header name in all lowercase letters (`blazor-enviornment`), you're welcome to use any casing that you desire. For example, a header name that capitalizes each word (`Blazor-Enviornment`) is supported.
```
```suggestion
> Although the Blazor framework issues the header name in all lowercase letters (`blazor-environment`), you're welcome to use any casing that you desire. For example, a header name that capitalizes each word (`Blazor-Environment`) is supported.
```

guardrex commented:
Going ahead with this. If there's additional feedback, I can create a new PR.
